File: management_api_app/models/domain/resource_template.py
from typing import Dict, Any, List

from pydantic import Field

from models.domain.azuretremodel import AzureTREModel
from models.domain.resource import ResourceType


class Property(AzureTREModel):
    type: str = Field(title="Property type")
    title: str = Field("", title="Property description")
    description: str = Field("", title="Property description")
    default: Any = Field(None, title="Default value for the property")


class ResourceTemplate(AzureTREModel):
    id: str
    name: str = Field(title="Unique template name")
    description: str = Field(title="Template description")
    version: str = Field(title="Template version")
    resourceType: ResourceType = Field(title="Type of resource this template is for (workspace/service)")
    current: bool = Field(title="Is this the current version of this template")
    type: str = "object"
    required: List[str] = Field(title="List of properties which must be provided")
    properties: Dict[str, Property] = Field(title="Template properties")

File: ui/app/src/models/resourceTemplate.ts
import { ResourceType } from "./resourceType";

export interface ResourceTemplate {
  id: string,
  name: string,
  type: string,
  description: string,
  version: string,
  title: string,
  resourceType: ResourceType,
  current: boolean,
  properties: any,
  allOf?: Array<any>,
  system_properties: any,
  actions: Array<TemplateAction>,
  customActions: Array<TemplateAction>,
  required: Array<string>,
  uiSchema: any,
  pipeline: any
}

export const sanitiseTemplateForRJSF = (template: ResourceTemplate) => {
  if (template.properties) {
    Object.keys(template.properties).forEach((key: string) => {
      Object.keys(template.properties[key]).forEach((name: string) => {
        if (template.properties[key][name] === null) {
          delete template.properties[key][name]
        }
      });
    });
  }

  const sanitised = {
    name: template.name,
    type: template.type,
    description: template.description,
    title: template.title,
    properties: template.properties,
    allOf: template.allOf,
    required: template.required,
    uiSchema: template.uiSchema
  }

  if (!sanitised.allOf) delete sanitised.allOf;

  return sanitised;
};

export interface TemplateAction {
  name: string,
  description: string
}

// make a sensible guess at an icon
export const getActionIcon = (actionName: string) => {
  switch(actionName.toLowerCase()){
      case 'start':
          return 'Play';
      case 'stop':
          return 'Stop';
      default:
          return 'Asterisk'
  }
};
